great_expectations/experimental/datasources/multi_model.py

```diff
@@ -0,0 +1,505 @@
+from __future__ import annotations
+
+import dataclasses
+import functools
+import itertools
+import logging
+from pprint import pformat as pf
+from typing import (
+    TYPE_CHECKING,
+    Dict,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Set,
+    Type,
+    cast,
+    overload,
+)
+
+import pydantic
+from typing_extensions import ClassVar, Literal, Protocol, runtime_checkable
+
+from great_expectations.core.batch_spec import SqlAlchemyDatasourceBatchSpec
+from great_expectations.execution_engine import ExecutionEngine
+from great_expectations.experimental.datasources.experimental_base_model import (
+    ExperimentalBaseModel,
+)
+from great_expectations.experimental.datasources.sources import _SourceFactories
+
+if TYPE_CHECKING:
+    from great_expectations.execution_engine import SqlAlchemyExecutionEngine
+    from great_expectations.experimental.datasources.interfaces import (
+        BatchRequest,
+        BatchRequestOptions,
+    )
+    from great_expectations.experimental.datasources.postgres_datasource import (
+        _DEFAULT_MONTH_RANGE,
+        _DEFAULT_YEAR_RANGE,
+        Batch,
+        BatchRequestError,
+        ColumnSplitter,
+    )
+
+# logging.basicConfig(level=logging.DEBUG)  # TODO: remove me
+LOGGER = logging.getLogger(__name__)
+
+
+class DataAssetConfig(ExperimentalBaseModel):
+    name: str
+    type: str
+
+    # non-field private attrs
+    _datasource: Datasource = pydantic.PrivateAttr()
+
+    @property
+    def datasource(self) -> Datasource:
+        return self._datasource
+
+    # TODO (kilo): remove setter and add custom init for DataAsset to inject datasource in constructor??
+    @datasource.setter
+    def datasource(self, ds: Datasource):
+        assert isinstance(ds, Datasource)
+        self._datasource = ds
+
+
+class DataAsset(Protocol):
+    name: str
+    datasource: Datasource
+    config: DataAssetConfig
+
+    def __init__(self, config: DataAssetConfig, datasource: Datasource) -> None:
+        if isinstance(config, DataAssetConfig):
+            self.config = config
+        else:
+            raise TypeError(
+                f"Expected {DataAssetConfig.__name__} got {type(config).__name__}"
+            )
+        self.datasource = datasource
+
+    def get_batch_request(self, options: Optional[BatchRequestOptions]) -> BatchRequest:
+        raise NotImplementedError
+
+
+class DatasourceConfig(
+    ExperimentalBaseModel,
+    # metaclass=SomeDifferentMetaclass,
```

Comment:
This has not been implemented yet but this new Metaclass would need to perform a similar role as the existing `Metadataclass`.
At minimum it would need to register all the `Config` models and map them with the correct `type: str`.
Example: `type = 'postgres'`